pkg/services/ngalert/state/cache.go

```diff
@@ -33,18 +32,42 @@ func newCache(logger log.Logger, metrics *metrics.State, externalURL *url.URL) *
 	}
 }
 
-func (c *cache) getOrCreate(ctx context.Context, alertRule *ngModels.AlertRule, result eval.Result) *State {
+func (c *cache) getOrCreate(ctx context.Context, alertRule *ngModels.AlertRule, result eval.Result, extraLabels data.Labels) *State {
 	c.mtxStates.Lock()
 	defer c.mtxStates.Unlock()
 
-	// clone the labels so we don't change eval.Result
-	labels := result.Instance.Copy()
-	attachRuleLabels(labels, alertRule)
-	ruleLabels, annotations := c.expandRuleLabelsAndAnnotations(ctx, alertRule, labels, result)
+	ruleLabels, annotations := c.expandRuleLabelsAndAnnotations(ctx, alertRule, result, extraLabels)
 
-	// if duplicate labels exist, alertRule label will take precedence
-	lbs := mergeLabels(ruleLabels, result.Instance)
-	attachRuleLabels(lbs, alertRule)
+	lbs := make(data.Labels, len(extraLabels)+len(ruleLabels)+len(result.Instance))
+	dupes := make(data.Labels)
```

Comment:
NIT: This section could be simplified and reuse code by modifying `mergeLabels` to return `dupes`. Then this section becomes something like:

```
// if duplicate labels exist, reserved label will take precedence
lbs, dupes := mergeLabels(extraLabels, ruleLabels)
if len(dupes) > 0 {
	c.log.Warn("rule declares one or many reserved labels. Those rules labels will be ignored", "labels", dupes)
}
// if duplicate labels exist, reserved or alert rule label will take precedence
lbs, dupes = mergeLabels(lbs, result.Instance)
if len(dupes) > 0 {
	c.log.Warn("evaluation result contains either reserved labels or labels declared in the rules. Those labels from the result will be ignored", "labels", dupes)
}
```

Alternatively (or in addition to), maybe this section that can be extracted into a separate function for ease of reading.

WDYT?